File: Server/src/main/scala/jk_5/nailed/server/command/CommandSudo.java
package jk_5.nailed.server.command;

import java.util.*;

import com.google.common.base.*;

import net.minecraft.command.*;
import net.minecraft.server.*;

import cpw.mods.fml.common.*;

import jk_5.nailed.api.*;
import jk_5.nailed.api.map.Map;
import jk_5.nailed.api.player.*;

/**
 * No description given
 *
 * @author jk-5
 */
public class CommandSudo extends NailedCommand {

    public CommandSudo() {
        super("sudo");
    }

    @Override
    public void processCommandWithMap(ICommandSender sender, Map map, String[] args) {
        Player player = NailedAPI.getPlayerRegistry().getPlayerByUsername(args[0]);
        args[0] = null;
        String cmd = Joiner.on(" ").skipNulls().join(args);
        if(player != null){
            FMLCommonHandler.instance().getMinecraftServerInstance().getCommandManager().executeCommand(player.getEntity(), cmd);
        }else{
            throw new CommandException("Unknown player");
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public List<String> addAutocomplete(ICommandSender sender, String[] args) {
        if(args.length == 1){
            return getUsernameOptions(args);
        }else if(args.length == 2){
            Player target = NailedAPI.getPlayerRegistry().getPlayerByUsername(args[0]);
            return getOptions(args, (Iterable<String>) MinecraftServer.getServer().getCommandManager().getPossibleCommands(target.getEntity()));
        }else if(args.length > 2){
            Player target = NailedAPI.getPlayerRegistry().getPlayerByUsername(args[0]);
            ICommand cmd = (ICommand) MinecraftServer.getServer().getCommandManager().getCommands().get(args[1]);
            if(cmd == null){
                return null;
            }
            String[] newArgs = new String[args.length - 2];
            System.arraycopy(args, 2, newArgs, 0, args.length - 2);
            return (List<String>) cmd.addTabCompletionOptions(target.getEntity(), newArgs);
        }else{
            return null;
        }
    }
}

File: Server/src/main/scala/jk_5/nailed/server/command/CommandSudo.scala
package jk_5.nailed.server.command

import net.minecraft.command.{ICommand, ICommandSender}
import jk_5.nailed.api.map.Map
import jk_5.nailed.api.NailedAPI
import net.minecraft.server.MinecraftServer
import scala.collection.JavaConverters._

/**
 * No description given
 *
 * @author jk-5
 */
object CommandSudo extends ScalaCommand {

  val name = "sudo"
  val usage = "/sudo <player> <command> - Execute a command as another player"

  override def processCommandWithMap(sender: ICommandSender, map: Map, args: Array[String]) = Option(NailedAPI.getPlayerRegistry.getPlayerByUsername(args(0))) match {
    case Some(player) =>
      val newArray = new Array[String](args.length - 1)
      System.arraycopy(args, 1, newArray, 0, newArray.length)
      MinecraftServer.getServer.getCommandManager.executeCommand(player.getEntity, newArray.mkString(" "))
    case None => throw new CommandException("Unknown player")
  }

  override def addAutocomplete(sender: ICommandSender, args: Array[String]): List[String] = args.length match {
    case 1 => getUsernameOptions(args)
    case 2 =>
      val target = NailedAPI.getPlayerRegistry.getPlayerByUsername(args(0))
      getOptions(args, MinecraftServer.getServer.getCommandManager.getPossibleCommands(target.getEntity).asInstanceOf[java.util.List[String]] asScala)
    case l if l > 2 =>
      val target = NailedAPI.getPlayerRegistry.getPlayerByUsername(args(0))
      val cmd = MinecraftServer.getServer.getCommandManager.getCommands.get(args(1)).asInstanceOf[ICommand]
      if(cmd == null) null
      else{
        val newArgs = new Array[String](args.length - 2)
        System.arraycopy(args, 2, newArgs, 0, args.length - 2)
        cmd.addTabCompletionOptions(target.getEntity, newArgs).asInstanceOf[java.util.List[String]].asScala.toList
      }
    case _ => null
  }
}
